File: Tree_Java/Tree.java
//==========================================================================
// cs12xre                         Homework 7              Prashant Singh
//--------------------------------------------------------------------------
// File: Tree.java
//
// Description: Contains the TNode object and the member functions of
//              the Tree class.
//==========================================================================
public class Tree<UCSDStudent extends Base> extends Base {

	/* data fields */
	private TNode root;
	private long occupancy; 
	private String treeName;

	/* debug flag */
	private static boolean debug;

	/* debug messages */
	private static final String ALLOCATE = " - Allocating]\n";
	private static final String AND = " and ";
	private static final String COMPARE = " - Comparing ";
	private static final String INSERT = " - Inserting ";
	private static final String TREE = "[Tree ";


	/**
	 * This method allocates and initializes the memory
	 * associated with a Tree.
	 *
	 * @param  name   The name of the tree.
	 */
	public Tree (String name) {
		/* Debug Message*/
		if(debug) {
			System.err.println(TREE + name + ALLOCATE);
		}

		/* Initializing the tree */
		treeName = name;
		occupancy = 0;
		root = null;

	}

	/**
	 * This function sets the debug mode as off
	 */
	public static void debugOff () {

		debug = false;
	}

	/**
	 * This function sets the debug mode as on
	 */
	public static void debugOn () {

		debug = true;
	}

	/**
	 * Returns the tree's name
	 * @return name of the tree
	 */
	public String getName() {
		return treeName;
	}

	/**
	 * This method will insert the element in the Tree.
	 * If the element cannot be inserted, false will be returned.
	 * If the element can be inserted, the element is inserted
	 * and true is returned.
	 *
	 * @param   element       The element to insert.
	 * @return  true or false indicating success or failure of insertion
	 */
	public boolean insert (UCSDStudent element) {

		/* Insertion - Begins */
		TNode working = null;
		/* If root is null, insert as root Node */
		if(root == null) {
			/* Debug Message*/
			if(debug) {
				System.err.println(TREE + treeName + 
						INSERT + element.getName() + "]");
			}
			root = new TNode(element);
			root.parent = root;
			working = root;
		}/* else begins inserting somewhere down the tree */
		else {
			working = root;
			/* When child nodes are not null */
			while(working.right != null || working.left != null) {
				/* Debug Message*/
				if(debug) {
					System.err.println(TREE + treeName + 
							COMPARE + element.getName() + 
							AND + working.data.getName() + "]");
				}
				/* element greater than working - move right */
				if(element.isGreaterThan(working.data)) {              	
					if(working.right == null) {
						/* Debug Message*/
						if(debug) {
							System.err.println(TREE + treeName + 
									INSERT + element.getName() + "]");
						}
						working.right = new TNode(element);
						working.right.parent = working;
						return true;
					}
					else {
						/* setting height while traversing */
						working.height += 1;
						working = working.right;
					}
				}/* else traverse left */
				else {
					if(working.left == null) {
						/* Debug Message*/
						if(debug) {
							System.err.println(TREE + treeName + 
									INSERT + element.getName() + "]");
						}
						working.left = new TNode(element);
						working.left.parent = working;
						return true;
					}
					else
						/* setting height while traversing */
						working.height += 1;
					working = working.left;
				}
			}

			/* When child nodes are null */
			if(working.data.isGreaterThan(element)) {
				/* Debug Message*/
				if(debug) {
					System.err.println(TREE + treeName + 
							COMPARE + element.getName() + 
							AND + working.data.getName() + "]");
					System.err.println(TREE + treeName + 
							INSERT + element.getName() + "]");
				}
				/* setting height while inserting */
				working.height += 1;
				working.left = new TNode(element);
				working.left.parent = working;
			}
			else {
				/* Debug Message*/
				if(debug) {
					System.err.println(TREE + treeName + 
							COMPARE + element.getName() + 
							AND + working.data.getName() + "]");
					System.err.println(TREE + treeName + 
							INSERT + element.getName() + "]");
				}
				/* setting height while inserting */
				working.height += 1;
				working.right = new TNode(element);
				working.right.parent = working;
			}
		}
		/* Insertion - Ends */

		/* Calculating the balance while moving up the tree - Begins */
		while(working!=root) {
			/* setting balance of working node */
			working.balance = (working.left==null? -1:working.left.height)
					- (working.right==null ? -1:working.right.height);
			/* setting balance of working's parent node */
			working.parent.balance = (working.parent.left==null? 
					-1:working.parent.left.height)
					- (working.parent.right==null ? 
							-1:working.parent.right.height);
			/* Traversing up the tree */
			working = working.parent;
		}
		/* Calculating balance - Ends */

		return true;
	}

	/**
	 * This method will locate the location in the
	 * tree for the insert or lookup.
	 *
	 * @param   element  The element needing a location.
	 * @return  element if item found, or NULL if not found
	 */
	public UCSDStudent lookup (UCSDStudent element) {

		/* Look Up - Begins */
		TNode working = null;
		/* If element is root */
		if(root != null && root.data.equals(element)) {
			if(!root.hasBeenDeleted)
				return root.data;
			else
				return null;
		}/* else find element somewhere down the tree */
		else {
			working = root;
			/* When child nodes are not null */
			while(working.right != null || working.left != null) {
				/* element equal to working */
				/* Debug Message*/
				if(debug) {
					System.err.println(TREE + treeName + 
							COMPARE + element.getName() + 
							AND + working.data.getName() + "]");
				}
				if(working.data.equals(element)) {
					/* check working if its deleted */
					if(!working.hasBeenDeleted)
						return working.data;
					else
						return null; // not there
				}/* else traverse right */
				else if(element.isGreaterThan(working.data)) {
					if(working.right != null)
						working = working.right;
					else
						return null; // not there
				}/* else traverse left */
				else {
					if(working.left != null)
						working = working.left;
					else
						return null; // not there
				}
			}

			/* When child nodes are null */
			/* Debug Message*/
			if(debug) {
				System.err.println(TREE + treeName + 
						COMPARE + element.getName() + 
						AND + working.data.getName() + "]");
			}
			if(working.data.equals(element)) {
				/* check working if its deleted */
				if(!working.hasBeenDeleted)
					return working.data;
				else
					return null; // not there
			}
		}
		/* Look Up - Ends */

		return null;    // not there
	}

	/**
	 * This method will remove the element in the hash table.  If
	 * removed, the element is returned.  If the element
	 * is not found, NULL will be returned to the user.
	 *
	 * @param   element  The element to look up.
	 * @return  element if found, else NULL
	 */
	public UCSDStudent remove (UCSDStudent element) {

		/* Remove - Begins */
		TNode working = null;
		/* If element is root & is null */
		if(root == null) {
			return null;
		}/* If element is root & is not null */
		else if(root != null && root.data.equals(element)) {
			if(!root.hasBeenDeleted) {
				root.hasBeenDeleted = true;
				occupancy--;
				return root.data;
			}
			else
				return null;
		}/* else find element somewhere down the tree */
		else {
			working = root;
			/* When child nodes are not null */
			while(working.right != null || working.left != null) {
				/* element equal to working */
				/* Debug Message*/
				if(debug) {
					System.err.println(TREE + treeName + 
							COMPARE + element.getName() + 
							AND + working.data.getName() + "]");
				}
				if(working.data.equals(element)) {
					/* check working if its deleted */
					if(!working.hasBeenDeleted) {
						working.hasBeenDeleted = true;
						occupancy--;
						return working.data;
					}
					else
						return null; // not there
				}/* else traverse right */
				else if(element.isGreaterThan(working.data)) {
					if(working.right != null)
						working = working.right;
					else
						return null; // not there
				}/* else traverse left */
				else {
					if(working.left != null)
						working = working.left;
					else
						return null; // not there
				}
			}

			/* When child nodes are null */
			/* Debug Message*/
			if(debug) {
				System.err.println(TREE + treeName + 
						COMPARE + element.getName() + 
						AND + working.data.getName() + "]");
			}
			if(working.data.equals(element)) {
				/* check working if its deleted */
				if(!working.hasBeenDeleted) {
					working.hasBeenDeleted = true;
					occupancy--;
					return working.data;
				}
				else
					return null; // not there
			}
		}
		/* Remove - Ends */

		return null;    // not there
	}

	/**
	 * Creates a string representation of this tree. This method first
	 * adds the general information of this tree, then calls the
	 * recursive TNode function to add all nodes to the return string 
	 *
	 * @return  String representation of this tree 
	 */
	public String toString () {
		String string = "Tree " + treeName + ":\noccupancy is ";
		string += occupancy + " elements.";

		if(root != null)
			string += root.writeAllTNodes();

		return string;
	}

	private class TNode {

		public UCSDStudent data;
		public TNode left, right, parent;
		public boolean hasBeenDeleted;

		/* left child's height - right child's height */
		public long balance;
		/* 1 + height of tallest child, or 0 for leaf */
		public long height;

		/**
		 * This method allocates and initializes the memory
		 * associated with a TNode.
		 *
		 * @param  element   The data for the TNode.
		 */
		public TNode (UCSDStudent element) {

			/* Initializing TNode */
			data = element;
			left = null;
			right = null;
			parent = null;
			balance = 0;
			height = 0;
			hasBeenDeleted = false;
			occupancy++;
		}

		/**
		 * Creates a string representation of this node. Information
		 * to be printed includes this node's height, its balance,
		 * and the data its storing.
		 *
		 * @return  String representation of this node 
		 */

		public String toString () {
			return "at height:  " + height + "  with balance: " +
					balance + "  " + data;
		}

		/**
		 * Writes all TNodes to the String representation field. 
		 * This recursive method performs an in-order
		 * traversal of the entire tree to print all nodes in
		 * sorted order, as determined by the keys stored in each
		 * node. To print itself, the current node will append to
		 * tree's String field.
		 */
		public String writeAllTNodes () {
			String string = "";
			if (left != null)
				string += left.writeAllTNodes ();
			if (!hasBeenDeleted) 
				string += "\n" + this;          
			if (right != null)
				string += right.writeAllTNodes ();

			return string;
		}
	}
}

File: Tree_C/Tree.c
//==========================================================================
// cs12xre                         Homework 7              Prashant Singh
//--------------------------------------------------------------------------
// File: Tree.c
//
// Description: Contains the TNode structure and the member functions of
//              the Tree class.
//==========================================================================
#include <cstdlib>
#include <cstdio>
#include <iostream>
#include <string.h>
#include "Tree.h"
using namespace std;

#ifndef TRUE
#define TRUE 1
#endif

#ifndef FALSE
#define FALSE 0
#endif

static const char ALLOCATE[] = " - Allocating]\n";
static const char DEALLOCATE[] = " has been deallocated]\n";
static const char AND[] = " and ";
static const char COMPARE[] = " - Comparing ";
static const char INSERT[] = " - Inserting ";
static const char TREE[] = "[Tree ";

struct TNode {
	Base * data;
	TNode * left, * right, *parent;
	static long occupancy;
	unsigned long hasBeenDeleted;

	// left child's height - right child's height
	long balance;

	// 1 + height of tallest child, or 0 for leaf
	long height;

	/**
	 * This method allocates and initializes the memory
	 * associated with a TNode.
	 *
	 * @param  element   The element for data of the TNode.
	 */
	TNode (Base * element) : data (element), left (0), right (0),
			parent (0), hasBeenDeleted (FALSE), balance (0), height(0) {

		/* Initializing TNode */
		data = element;
		left = NULL;
		right = NULL;
		parent = NULL;
		balance = 0;
		height = 0;
		hasBeenDeleted = FALSE;
		occupancy++;

	}

	/**
	 * This method deallocates the memory
	 * associated with a TNode.
	 *
	 * @param  void
	 */
	~TNode (void) {
		// Deallocating the memory
		delete(data);
		delete(left);
		delete(right);
		delete(parent);
	}

	ostream & Write (ostream & stream) const {
		stream << "at height:  " << height << "  with balance:  "
				<< balance << "  ";
		return data->Write (stream) << "\n";
	}
	ostream & Write_AllTNodes (ostream & stream) const {
		if (left)
			left->Write_AllTNodes (stream);
		if (!hasBeenDeleted)
			Write (stream);
		if (right)
			right->Write_AllTNodes (stream);

		return stream;
	}

	/**
	 * This method deallocates the memory
	 * associated with all TNodes in the Tree
	 * recursively.
	 *
	 * @param  void.
	 */
	void delete_AllTNodes (void) {
		// Deleting the nodes recursively
		if (left)
			left->delete_AllTNodes ();
		if (right)
			right->delete_AllTNodes ();

		if(this != NULL)
			delete (this);
	}
};

// initialization of static data fields of Tree
long TNode :: occupancy = 0;
bool Tree :: debug_on = 0;

/***************************************************************************
% Routine Name : Tree :: Set_Debug  (public)
% File :         Tree.c
%
% Description :  Sets the debug option as on or off.
***************************************************************************/
void Tree :: Set_Debug (bool option) {

	debug_on = option;
}

Tree :: Tree (const char * name) : root (0), tree_name (strdup (name))
/***************************************************************************
% Routine Name : Tree :: Tree  (public)
% File :         Tree.c
% 
% Description :  Initializes Tree, root pointer to NULL, and occupancy to 0.
***************************************************************************/

{
	/* Debug Message*/
	if(debug_on) {
		cerr << TREE << name << ALLOCATE << endl;
	}

	/* Initializing the tree */
	tree_name = name;
	root = NULL;

}

Tree :: ~Tree (void)
/***************************************************************************
% Routine Name : Tree :: ~Tree  (public)
% File :         Tree.c
% 
% Description :  deallocates memory associated with the Tree.  It
%                will also delete all the memory of the elements within
%                the table.
***************************************************************************/

{
	// freeing the memory
	free( (void *)tree_name);
}	/* end: ~Tree */

/***************************************************************************
% Routine Name : Tree :: Insert  (public)
% File :         Tree.c
%
% Description :  Inserts the element into the binary tree. Returns true or
%				 false indicating success of insertion.
%
% Parameters descriptions :
%
% name               description
% ------------------ ------------------------------------------------------
% element            A reference to the element to insert in the tree.
% <return>           1(TRUE) if success or 0(FALSE) if failed to insert.
***************************************************************************/
unsigned long Tree :: Insert (Base * element) {

	/* Insertion - Begins */
	TNode* working = NULL;
	/* If root is NULL, insert as root Node */
	if(root == NULL) {
		/* Debug Message*/
		if(debug_on) {
			cerr<<TREE << tree_name <<
					INSERT << (const char *) * element << "]" << endl;
		}
		root = new TNode(element);
		root->parent = root;
		working = root;
	}/* else begins inserting somewhere down the tree */
	else {
		working = root;
		/* When child nodes are not NULL */
		while(working->right != NULL || working->left != NULL) {
			/* Debug Message*/
			if(debug_on) {
				cerr<<TREE << tree_name <<
						COMPARE << (const char *) * element <<
						AND << (const char *) * (working->data)<< "]" << endl;
			}
			/* element greater than working - move right */
			if(*element > *(working->data)) {
				if(working->right == NULL) {
					/* Debug Message*/
					if(debug_on) {
						cerr<<TREE << tree_name <<
								INSERT << (const char *) * element << "]" << endl;
					}
					working->right = new TNode(element);
					working->right->parent = working;
					return TRUE;
				}
				else {
					/* setting height while traversing */
					working->height += 1;
					working = working->right;
				}
			}/* else traverse left */
			else {
				if(working->left == NULL) {
					/* Debug Message*/
					if(debug_on) {
						cerr<<TREE << tree_name <<
								INSERT << (const char *) * element << "]" << endl;
					}
					working->left = new TNode(element);
					working->left->parent = working;
					return TRUE;
				}
				else
					/* setting height while traversing */
					working->height += 1;
				working = working->left;
			}
		}

		/* When child nodes are NULL */
		if(*(working->data) > *element) {
			/* Debug Message*/
			if(debug_on) {
				cerr<<TREE << tree_name <<
						COMPARE << (const char *) * element <<
						AND <<(const char *) * (working->data) << "]" << endl;
				cerr<<TREE << tree_name <<
						INSERT << (const char *) * element << "]" << endl;
			}
			/* setting height while inserting */
			working->height += 1;
			working->left = new TNode(element);
			working->left->parent = working;
		}
		else {
			/* Debug Message*/
			if(debug_on) {
				cerr<<TREE << tree_name <<
						COMPARE << (const char *) * element <<
						AND << (const char *) * (working->data) << "]" << endl;
				cerr<<TREE << tree_name <<
						INSERT << (const char *) * element << "]" << endl;
			}
			/* setting height while inserting */
			working->height += 1;
			working->right = new TNode(element);
			working->right->parent = working;
		}
	}
	/* Insertion - Ends */

	/* Calculating the balance while moving up the tree - Begins */
	while(working!=root) {
		/* setting balance of working node */
		working->balance = (working->left==NULL? -1:working->left->height)
	    							- (working->right==NULL ? -1:working->right->height);
		/* setting balance of working's parent node */
		working->parent->balance = (working->parent->left==NULL?
				-1:working->parent->left->height)
	    							- (working->parent->right==NULL ?
	    									-1:working->parent->right->height);
		/* Traversing up the tree */
		working = working->parent;
	}
	/* Calculating balance - Ends */

	return TRUE;
}

/***************************************************************************
% Routine Name : Tree :: Lookup  (public)
% File :         Tree.c
%
% Description :  Looks up the matching data in the binary tree. Returns a
%				 pointer to the data if found, null otherwise.
%
% Parameters descriptions :
%
% name               description
% ------------------ ------------------------------------------------------
% element            A reference to the element to lookup.
% <return>           A reference to the data found or null if not.
***************************************************************************/
const Base * Tree :: Lookup (const Base * element) const {

	/* Look Up - Begins */
	TNode* working = NULL;
	/* If element is root */
	if(root != NULL && root->data == element) {
		if(!root->hasBeenDeleted)
			return root->data;
		else
			return NULL;
	}/* else find element somewhere down the tree */
	else {
		working = root;
		/* When child nodes are not NULL */
		while(working->right != NULL || working->left != NULL) {
			/* element equal to working */
			/* Debug Message*/
			if(debug_on) {
				cerr << TREE << tree_name <<
						COMPARE << (const char *) * element <<
						AND << (const char *) * (working->data) << "]"<< endl;
			}
			if(*(working->data) == *element) {
				/* check working if its deleted */
				if(!working->hasBeenDeleted)
					return working->data;
				else
					return NULL; // not there
			}/* else traverse right */
			else if(*element > *(working->data)) {
				if(working->right != NULL)
					working = working->right;
				else
					return NULL; // not there
			}/* else traverse left */
			else {
				if(working->left != NULL)
					working = working->left;
				else
					return NULL; // not there
			}
		}

		/* When child nodes are null */
		/* Debug Message*/
		if(debug_on) {
			cerr << TREE << tree_name <<
					COMPARE << (const char *) * element <<
					AND << (const char *) * (working->data) << "]"<< endl;
		}
		if(*(working->data) == *element) {
			/* check working if its deleted */
			if(!working->hasBeenDeleted)
				return working->data;
			else
				return NULL; // not there
		}
	}
	/* Look Up - Ends */
	return NULL;
}

/***************************************************************************
% Routine Name : Tree :: Remove  (public)
% File :         Tree.c
%
% Description :  Removes the matching data from the binary tree. Returns a
%				 pointer to the data if found, null otherwise.
%
% Parameters descriptions :
%
% name               description
% ------------------ ------------------------------------------------------
% element            A reference to the element to remove in the tree.
% <return>           A reference to the element removed or null if failed
%					 to removed.
***************************************************************************/
Base * Tree :: Remove (const Base * element) {
	/* Remove - Begins */
	TNode* working = NULL;
	/* If element is root & is NULL */
	if(root == NULL) {
		return NULL;
	}/* If element is root & is not NULL */
	else if(root != NULL && root->data == element) {
		if(!root->hasBeenDeleted) {
			root->hasBeenDeleted = TRUE;
			return root->data;
		}
		else
			return NULL;
	}/* else find element somewhere down the tree */
	else {
		working = root;
		/* When child nodes are not NULL */
		while(working->right != NULL || working->left != NULL) {
			/* element equal to working */
			/* Debug Message*/
			if(debug_on) {
				cerr << TREE << tree_name <<
						COMPARE << (const char *) * element <<
						AND << (const char *) * (working->data) << "]"<< endl;
			}
			if(*(working->data) == *element) {
				/* check working if its deleted */
				if(!working->hasBeenDeleted) {
					working->hasBeenDeleted = TRUE;
					return working->data;
				}
				else
					return NULL; // not there
			}/* else traverse right */
			else if(*element > *(working->data)) {
				if(working->right != NULL)
					working = working->right;
				else
					return NULL; // not there
			}/* else traverse left */
			else {
				if(working->left != NULL)
					working = working->left;
				else
					return NULL; // not there
			}
		}

		/* When child nodes are NULL */
		/* Debug Message*/
		if(debug_on) {
			cerr << TREE << tree_name <<
					COMPARE << (const char *) * element <<
					AND << (const char *) * (working->data) << "]"<< endl;
		}
		if(*(working->data) == *element) {
			/* check working if its deleted */
			if(!working->hasBeenDeleted) {
				working->hasBeenDeleted = TRUE;
				return working->data;
			}
			else
				return NULL; // not there
		}
	}
	/* Remove - Ends */

	return NULL;
}

ostream & Tree :: Write (ostream & stream) const
/***************************************************************************
% Routine Name : Tree :: Write (public)
% File :         Tree.c
% 
% Description : This function will output the contents of the Tree table
%               to the stream specificed by the caller.  The stream could be
%               cerr, cout, or any other valid stream.
%
% Parameters descriptions :
% 
% name               description
% ------------------ ------------------------------------------------------
% stream             A reference to the output stream.
% <return>           A reference to the output stream.
***************************************************************************/

{
        stream << "Tree " << tree_name << ":\n"
		<< "occupancy is " << TNode :: occupancy << " elements.\n";

	return (root) ? root->Write_AllTNodes (stream) : stream;
}       /* end: Write */

